app/resources/api/v2/asset_audit_resource.rb

```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
+
+module Api
+  module V2
+    # Provides a JSON API representation of AssetAudit
+    # See: http://jsonapi-resources.com/ for JSONAPI::Resource documentation
+    class AssetAuditResource < BaseResource
+      model_name 'AssetAudit'
+
+      # @!attribute key
+      #  @return [String] the key of the asset audit.
+      attribute :key
+
+      # @!attribute message
+      #  @return [String] the message of the asset audit.
+      attribute :message
+
+      # @!attribute created_by
+      #  @return [String] the user who created the asset audit.
+      attribute :created_by
+
+      # @!attribute asset_uuid
+      #  @return [String] the uuid of the asset associated with the asset audit.
+      attribute :asset_uuid
+
+      # @!attribute witnessed_by
+      #  @return [String] the user who witnessed the asset audit.
+      attribute :witnessed_by
+
+      # @!attribute metadata
+      #  @return [Hash] the metadata of the asset audit.
+      attribute :metadata  # Currently known clients (AssetAudits app) are sending null; unsure of the expected format.
+
+      # Sets the Asset on the model using the UUID provided in the API create/update request.
+      #
+      # @param name [String] the uuid of the associated asset.
+      # @return [void]
+      def asset_uuid=(uuid)
+        @model.asset = Uuid.with_external_id(uuid).include_resource.map(&:resource).first
+      end
+
+      # Transforms the Asset into its UUID when generating an API query response.
+      #
+      # @return [String] the uuid of the associated asset.
+      def asset_uuid
+        @model.asset.uuid
+      end
+
+      # Gets the list of fields which are updatable on an existing AssetAudit.
+      #
+      # @param _context [JSONAPI::Resource::Context] not used.
+      # @return [Array<Symbol>] the list of updatable fields.
+      def updatable_fields(_context)
```

Comment:
Should we fix other classes that don't include `self.`?

Reply:
SS in general.  I agree and it probably needs to be a separate cleanup task.  The intention is there in other models, but it won't be doing what the developer intended.  We either need to remove the definition or accept it might break some use cases (by trying to correct the code).  See, for example:  https://github.com/sanger/sequencescape/blob/develop/app/resources/api/v2/plate_purpose_resource.rb#L80-L82

Reply:
In fact that's the only example, so perhaps ask @yoldas what he had been trying to achieve before we create a whole story out of it.

Reply:
Yes, I think updatable_fields should be self.updatable_fields as you found from the API docs. It is to prevent modification of existing purposes. I have migrated some v1 resources to v2 for Limber, this may be one of them.

Reply:
Do you know who/which apps use the plate purpose resource?  If they are currently using it to update existing ones (because it doesn't protect against that) and we fix it, then that will break something somewhere else.

Reply:
Limber config:generate task uses it. I have added it to use API V2 to allow asset shape from purpose config so I can specify the Chromium Chip asset shape. V1 did not allow passing shape. It is the direct translation from v1 to v2. Specifically, it is used here: https://github.com/sanger/limber/blob/develop/lib/purpose_config.rb#L103

Reply:
Purposes are not supposed to be updated using the API. It is a dangerous thing to do. We edit the DB for changing if we have to.

Reply:
Personally I would prefer the Limber config for a purpose be the source of truth so config:generate should update the database for purposes. Had several issues in the past where flags set previously on purposes were not updated as a pipeline got developed and then broke on deployment as we forget to check and manually update the database. But agree dangerous to allow generally, would need to restrict it to only config:generate on deployment.

Reply:
Hmm, that's going to be tricky.  The only things I can think of are:

* Make the fields updatable and accept that anyone with access to the API could update them.
* Add a different endpoint that Limber uses to make these changes -- not great as it still leaves an open endpoint for such changes.
* Have some way of changing the return value of this method under some circumstances:
  * Perhaps the `config:generate` could use a different API key and perhaps that key could be identified in this method.  Kind of like a super-user flag on the API key config. (I've not looked into this)
  * Change Sequencescape to a temporary mode using a feature flag which allows all fields to be updated.  I was told `flipper` has a UI -- never seen it, but perhaps that could be used?
  * Something even more exotic, like allowing changes at a certain time of day, allowing it for 5 minutes after SS was last deployed or allowing the change if the request was made three times in a short period.  I don't like these options.

I think using a different API key is probably the most appropriate option to balance convenience and security, but it's not the simplest to implement.

Reply:
I like the different API key idea too, deployments could use it.
Leave it as you've got it for now and we can discuss it in a retro or something.